In [25]:
# Load packages
from costs_benefits_ssp.cb_calculate import CostBenefits
import pandas as pd
import os
import pathlib
import numpy as np


In [26]:
# Define paths
SSP_PATH = pathlib.Path(os.getcwd()).parent.parent.as_posix()
SSP_RUN  = os.path.join(SSP_PATH, "ssp_modeling/ssp_run_output/sisepuede_summary_results_run_sisepuede_run_2025-10-29T19;49;25.722413")


CB_DEFAULT_DEFINITION_PATH = os.path.join(SSP_PATH,"ssp_modeling/cb/cb_cost_factors")
CB_DEFAULT_DEFINITION_FILE_PATH = os.path.join(CB_DEFAULT_DEFINITION_PATH, "cb_config_params.xlsx")

CB_OUTPUT = os.path.join(SSP_PATH,"ssp_modeling/cb/cb_results")

In [27]:
# Load data
ssp_data = pd.read_csv(os.path.join(SSP_RUN, "uganda.csv"))
att_primary = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_PRIMARY.csv"))
att_strategy = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_STRATEGY.csv"))
strategy_code_base = "BASE"

In [28]:
att_primary

,primary_id,design_id,strategy_id,future_id
0,0,0,0,0
1,70070,0,6004,0
2,74074,0,6008,0
3,75075,0,6009,0


In [29]:
# Instantiate CostBenefits object
cb = CostBenefits(ssp_data, att_primary, att_strategy, strategy_code_base)

# The export_db_to_excel method saves the initial configuration of the cost tables to an excel file.
# Each sheet represents a table in the cost and benefit program database.
# If the Excel file name is not given, the file will be saved with the default name cb_config_params.xlsx on the current python session.

#cb.export_db_to_excel(CB_DEFAULT_DEFINITION_FILE_PATH)

# Once that the excel file has been updated, we can reload it in order to update the cost factors database
cb.load_cb_parameters(CB_DEFAULT_DEFINITION_FILE_PATH)

The TX TX:FRST:INCREASE_SEQUESTRATION_NZ is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NZ is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NZ is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NZ is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_2 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_2 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_2 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_2 is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
Cargamos configuración de archivo excel
Se actualizó la base de datos


In [30]:
# Compute System Costs
results_system = cb.compute_system_cost_for_all_strategies()


************************************
*Strategy : PFLO:NZ (0/3)
************************************

---------Costs for: cb:wali:technical_cost:sanitation:unimp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:unimp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:omit_rural.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:capex.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:opex.
La variable

In [31]:
# Compute Technical Costs
results_tx = cb.compute_technical_cost_for_all_strategies()


************************************
*Strategy : PFLO:NZ (0/3)
************************************

---------Costs for: cb:inen:technical_cost:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:efficiency:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:tech

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:856: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_variable"] = cb_orm.diff_var
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:857: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_value"] = data_merged["difference"]
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/cos

---------Costs for: cb:ippu:technical_cost:abating_gases:other_fgases.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:flaring:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:leaks:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:technical_cost:consumer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:consumer_savings:consumer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ccsq:technical_cost:direct_air_capture:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:lvst:technical_cost:ent_ferm_mgmt:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical_cost:rice_mgmt:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical_cost:producer_food_waste:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:agrc:technical

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_savings:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:hdv_vehicles.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_savings:electrification:hdv_vehicles.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:hdv_charger.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:efficiency:non_electric.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:efficiency:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_cost:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalúa en

/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


In [32]:
# Combine results
results_all = pd.concat([results_system, results_tx], ignore_index = True)

#-------------POST PROCESS SIMULATION RESULTS---------------
# Post process interactions among strategies that affect the same variables
results_all_pp = cb.cb_process_interactions(results_all)

# SHIFT any stray costs incurred from 2015 to 2025 to 2025 and 2035
results_all_pp_shifted = cb.cb_shift_costs(results_all_pp)

Resolving Interactions in SCOE : TX:SCOE:INC_EFFICIENCY_APPLIANCE, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 


/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:660: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["variable"] = res_pre2025["variable"] + "_shifted" + (res_pre2025["time_period"]+SSP_GLOBAL_TIME_PERIOD_0).astype(str)#create a new variable so they can be recognized as shifted costs
/Users/fabianfuentes/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:661: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-

In [33]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:NZ,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:NZ,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:NZ,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:NZ,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:NZ,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [34]:
ssp_data.head()

,primary_id,region,time_period,area_agrc_crops_bevs_and_spices,area_agrc_crops_cereals,area_agrc_crops_fibers,area_agrc_crops_fruits,area_agrc_crops_herbs_and_other_perennial_crops,area_agrc_crops_nuts,area_agrc_crops_other_annual,...,yf_agrc_herbs_and_other_perennial_crops_tonne_ha,yf_agrc_nuts_tonne_ha,yf_agrc_other_annual_tonne_ha,yf_agrc_other_woody_perennial_tonne_ha,yf_agrc_pulses_tonne_ha,yf_agrc_rice_tonne_ha,yf_agrc_sugar_cane_tonne_ha,yf_agrc_tubers_tonne_ha,yf_agrc_vegetables_and_vines_tonne_ha,yf_lndu_supremum_pastures_tonne_per_ha
0,0,uganda,4,634960.593061,2.025945e+06,156073.139455,1.131815e+06,505.429083,513832.892470,2.915383e+06,...,2.172065,0.458120,1.528325,0.3674,0.799814,2.970174,77.726965,4.805888,4.035314,92.81
1,0,uganda,5,640126.452982,2.042427e+06,157342.906406,1.141024e+06,509.541112,518013.291655,2.939102e+06,...,2.196794,0.611231,1.681739,0.3674,0.805568,3.032345,78.023706,4.440105,4.013004,92.81
2,0,uganda,6,642767.262368,2.050853e+06,157992.016628,1.145731e+06,511.643198,520150.329354,2.951227e+06,...,2.210866,0.841393,2.293907,0.3674,0.807398,3.080626,61.430414,4.911305,4.023409,92.81
3,0,uganda,7,645322.479649,2.059006e+06,158620.088334,1.150285e+06,513.677153,522218.102852,2.962959e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81
4,0,uganda,8,647798.760608,2.066907e+06,159228.757514,1.154699e+06,515.648274,524221.998246,2.974329e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81


# Indoor air pollution benef

Health co-benefit from reduced household air pollution

To monetize the health benefits of reducing indoor air pollution from the use of traditional biomass stoves, we drew on estimates from the BAR-HAP model (The Benefits of Action to Reduce Household Air Pollution; WHO, 2021).
The study reports cost-effectiveness ratios ranging from 674 – 869 USD per DALY avoided for clean cooking interventions in Nepal, meaning that each avoided disability-adjusted life year (DALY) costs approximately 700 USD.

To adapt this metric to Uganda, we scaled the cost-effectiveness ratio by the relative difference in GDP per capita between Nepal (~1,450 USD) and Uganda (~1,070 USD), yielding an adjusted range of ≈ 500–640 USD per DALY avoided.
Assuming that the transition from solid fuels to clean cooking solutions avoids about 0.1 DALY per person per year, the implied annual health benefit equals 50–64 USD per person.

Given an average household size of five, this corresponds to an estimated 250–320 USD per household per year in health-related economic benefits.

This adjusted value was used in the cost–benefit analysis as the annual household health co-benefit of eliminating exposure to indoor air pollution from firewood and other solid fuels in Uganda.

In [35]:
import pandas as pd
import numpy as np

# parameters
benefit_per_person = 320   # USD/person/year. 
years = np.arange(4, 56)

# ['primary_id', 'time_period', 'population_gnrl_total']

# baseline
baseline = ssp_data[ssp_data['primary_id'] == 0][['time_period', 'population_gnrl_total', 'frac_scoe_heat_energy_residential_solid_biomass']]
baseline = baseline.rename(columns={'population_gnrl_total': 'population_baseline'})
baseline = baseline.rename(columns={'frac_scoe_heat_energy_residential_solid_biomass': 'frac_scoe_heat_energy_residential_solid_biomass_baseline'})

baseline.head()

,time_period,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline
0,4,39853824,0.960324
1,5,40996690,0.932779
2,6,42172327,0.901015
3,7,43381679,0.901015
4,8,44625710,0.901015


In [36]:
# baseline merge
merged = ssp_data[ssp_data['primary_id'] != 0][['primary_id', 'time_period', 'population_gnrl_total', 
                                                'frac_scoe_heat_energy_residential_solid_biomass']].merge(baseline, 
                                                                                                          on='time_period', how='left')
merged.tail(50)

,primary_id,time_period,population_gnrl_total,frac_scoe_heat_energy_residential_solid_biomass,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline
106,75075,6,42172327,0.901015,42172327,0.901015
107,75075,7,43381679,0.901015,43381679,0.901015
108,75075,8,44625710,0.901015,44625710,0.901015
109,75075,9,45905416,0.901015,45905416,0.901015
110,75075,10,47221819,0.901015,47221819,0.901015
111,75075,11,48562777,0.901015,48562777,0.901015
112,75075,12,49928247,0.881160,49928247,0.892538
113,75075,13,51318160,0.850218,51318160,0.879923
114,75075,14,52732427,0.803123,52732427,0.861368
115,75075,15,54170936,0.734175,54170936,0.834544


In [37]:
# frac_scoe_heat_energy_residential_solid_biomass 
merged['pop_diff'] = (merged['frac_scoe_heat_energy_residential_solid_biomass_baseline'] - merged['frac_scoe_heat_energy_residential_solid_biomass'] ) * merged['population_gnrl_total']

In [38]:
# difference calculation
merged['value'] =  merged['pop_diff'] * benefit_per_person

In [39]:
merged.head(30)

,primary_id,time_period,population_gnrl_total,frac_scoe_heat_energy_residential_solid_biomass,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline,pop_diff,value
0,70070,4,39853824,0.960324,39853824,0.960324,0.000000e+00,0.000000e+00
1,70070,5,40996690,0.932779,40996690,0.932779,0.000000e+00,0.000000e+00
2,70070,6,42172327,0.901015,42172327,0.901015,0.000000e+00,0.000000e+00
3,70070,7,43381679,0.901015,43381679,0.901015,0.000000e+00,0.000000e+00
4,70070,8,44625710,0.901015,44625710,0.901015,0.000000e+00,0.000000e+00
5,70070,9,45905416,0.901015,45905416,0.901015,0.000000e+00,0.000000e+00
6,70070,10,47221819,0.901015,47221819,0.901015,0.000000e+00,0.000000e+00
7,70070,11,48562777,0.901015,48562777,0.901015,0.000000e+00,0.000000e+00
8,70070,12,49928247,0.892538,49928247,0.892538,0.000000e+00,0.000000e+00
9,70070,13,51318160,0.859459,51318160,0.879923,1.050140e+06,3.360448e+08


In [40]:
merged = merged.merge(att_primary[['primary_id', 'strategy_id', 'future_id']], on='primary_id', how='left')
merged = merged.merge(att_strategy[['strategy_id', 'strategy_code']], on='strategy_id', how='left')

In [41]:
merged.to_csv(os.path.join(CB_OUTPUT, "cba_indoor_air_pollution.csv"), index = False)

In [42]:
merged['region']='uganda'
merged['difference_variable'] ='frac_scoe_heat_energy_residential_solid_biomass'
merged['difference_value'] = 'pop_diff'
merged['variable'] ='cb:scoe:indoor_air_pollution'

In [43]:
merged = merged[['strategy_code', 'future_id', 'region', 'time_period', 'difference_variable', 'value','variable']]
merged.head()

,strategy_code,future_id,region,time_period,difference_variable,value,variable
0,PFLO:NZ,0,uganda,4,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
1,PFLO:NZ,0,uganda,5,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
2,PFLO:NZ,0,uganda,6,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
3,PFLO:NZ,0,uganda,7,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
4,PFLO:NZ,0,uganda,8,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution


In [44]:
merged.to_csv(os.path.join(CB_OUTPUT, "cba_indoor_air_pollution.csv"), index = False)

In [45]:
results_all_pp_shifted= pd.concat([results_all_pp_shifted, merged], ignore_index = True)

In [46]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:NZ,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:NZ,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:NZ,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:NZ,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:NZ,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [47]:
results_all_pp_shifted.tail()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
37535,PFLO:NDC_2_2025_TUNED,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.229103e+08
37536,PFLO:NDC_2_2025_TUNED,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.275342e+08
37537,PFLO:NDC_2_2025_TUNED,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.321311e+08
37538,PFLO:NDC_2_2025_TUNED,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.366974e+08
37539,PFLO:NDC_2_2025_TUNED,0,uganda,55,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.412520e+08


In [48]:
# Save the results
results_all_pp_shifted.to_csv(os.path.join(CB_OUTPUT, "cba_results.csv"), index = False)